File: Engine/src/jvmMain/java/org/etieskrill/engine/input/controller/KeyCharacterTranslationController.java
package org.etieskrill.engine.input.controller;

import org.etieskrill.engine.graphics.camera.Camera;
import org.joml.Vector3f;

import static java.lang.Math.toRadians;

public class KeyCharacterTranslationController extends KeyCharacterController<Vector3f> {

    public KeyCharacterTranslationController(
            Vector3f translation,
            Camera camera
    ) {
        this(translation, camera, true);
    }

    public KeyCharacterTranslationController(
            Vector3f translation,
            Camera camera,
            boolean fixUpDirection
    ) {
        super(translation, (delta, target, deltaPosition, speed) -> {
            if (!deltaPosition.equals(0, 0, 0)) {
                deltaPosition.normalize();
            }
            if (fixUpDirection) {
                deltaPosition.x = -deltaPosition.x;
                deltaPosition.rotateY((float) toRadians(camera.getYaw()));
            } else {
                deltaPosition = camera.relativeTranslation(deltaPosition);
            }
            translation.set(deltaPosition.mul(speed));
        });
    }

}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/input/controller/KeyCharacterTranslationController.kt
package org.etieskrill.engine.input.controller

import org.etieskrill.engine.graphics.camera.Camera
import org.joml.Math.toRadians
import org.joml.Vector3f
import org.joml.times

class KeyCharacterTranslationController(
    translation: Vector3f,
    camera: Camera,
    fixUpDirection: Boolean = true
) : KeyCharacterController<Vector3f>(
    translation, { delta, target, deltaPosition, speed ->
        if (!deltaPosition.equals(0f, 0f, 0f)) {
            deltaPosition.normalize()
        }

        val deltaTranslation = if (fixUpDirection) {
            deltaPosition.x = -deltaPosition.x
            deltaPosition.rotateY(toRadians(camera.yaw))
        } else {
            camera.relativeTranslation(deltaPosition)
        }

        translation.set(deltaTranslation * speed)
    }
)
